nemo/core/optim/distributed_adam.py

```diff
@@ -318,3 +325,24 @@ def load_state_dict(self, state_dict):
                 old_main_param.copy_(new_main_param.detach())
             del state_dict['fp32_optim_fp32_params']
         return super().load_state_dict(state_dict)
+
+    def sharded_state_dict(self, model_sharded_state_dict):
+        optimizer_state_dict = self.state_dict()
```

Comment:
@ericharper With the new distributed optimizer state dict v2 which gathers all tensors on CPU, this line takes 10-20 seconds for the GPT-175B. Currently we call it once for loading and 4 times for saving (2 checkpoints (regular+last), each calls optimizer.sharded_state_dict() twice), but we actually need it only once for saving, so we should come up with a way to avoid calling this abundantly:
- for loading, we could have a `DistributedOptimizer.state_dict` alternative that doesn't gather the data, just builds the proper state dict structure
- for saving, we just need to organize the saving logic and be cautious about calling state_dict too many times